src/error.rs

```diff
@@ -85,14 +62,110 @@ pub(crate) enum JniErrorInternal {
+/// a Java exception was thrown. If we waited to materialize errors until exiting `with` or in a `catching()` block,
+/// the execute call would return an internal [`Error::Jni`] instead and the exception would be buried.
+///
+pub fn with_jni_env<'jvm, T>(
+    env: &mut JNIEnv<'jvm>,
+    f: impl FnOnce(&mut JNIEnv<'jvm>) -> jni::errors::Result<T>,
+) -> Result<'jvm, T> {
+    let result = f(env);
+    result.map_err(|e| match e {
+        jni::errors::Error::JavaException => {
+            let exception = match env.exception_occurred() {
+                Ok(ex) => ex,
+                Err(e) => return convert_non_throw_jni_error(e),
+            };
+            assert!(!exception.is_null());
+            if let Err(e) = env.exception_clear() {
+                return convert_non_throw_jni_error(e);
+            }
+            Error::Thrown(unsafe {
+                Local::from_jni(AutoLocal::new(JObject::from(exception), &env))
+            })
+        }
+        error => convert_non_throw_jni_error(error),
+    })
+}
+
+/// Plumbing utility to convert any [`jni::errors::Error`] that *isn't* from a thrown exception. This doesn't require
+/// a [`JNIEnv`] borrow and can be used in more contexts. However, it isn't implemented as an [`Into`] to avoid masking
+/// scenarios that do require an explicit exception check via [`with_jni_env()`].
+///
+/// # Panics
+///
+/// Panics if the error indicates a thrown Java exception.
+pub fn convert_non_throw_jni_error<T>(error: jni::errors::Error) -> Error<T> {
+    assert!(!matches!(error, jni::errors::Error::JavaException));
+    Error::from(JniError::from(JniErrorInternal::from(error)))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{
+        java::{self, lang::ThrowableExt, util::ArrayListExt},
+        prelude::*,
+    };
+
+    #[test]
+    fn with_jni_env_materializes_thrown_exceptions() {
+        Jvm::with(|jvm| {
+            let env = jvm.to_env();
+            // set exception state
+            env.throw_new("java/lang/Throwable", "some thrown exception")
+                .unwrap();
+            let err = with_jni_env(env, |_env| Err::<(), _>(jni::errors::Error::JavaException))
+                .unwrap_err();
+            let Error::Thrown(t) = err else { panic!("expected materialized exception"); };
+            assert_eq!(
+                "some thrown exception",
+                t.get_message().assert_not_null().into_rust(jvm).unwrap()
+            );
+            Ok(())
+        })
+        .unwrap();
+    }
+
+    #[test]
+    fn with_jni_env_maps_everything_else_to_jni() {
+        Jvm::with(|jvm| {
+            let err = with_jni_env(jvm.to_env(), |_env| {
+                Err::<(), _>(jni::errors::Error::TryLock)
+            })
+            .unwrap_err();
+            assert!(matches!(err, Error::Jni(_)));
+            Ok(())
+        })
+        .unwrap();
+    }
+
+    // XX this should likely move to an integration test suite
```

Comment:
Let me know if this suffices for now or if we want to tackle a more proper test suite first.